mgr/dashboard: RGW multisite sync flow

nmunet opened the pull request:
Fixes: https://tracker.ceph.com/issues/66915

Signed-off-by: Naman Munet <nmunet@redhat.com>


Sync Flow Demo - 

https://github.com/user-attachments/assets/be20ac03-c182-44c9-bb3d-1a9bd6825bc2

Directional Sync flow edit/delete is not possible yet as the core backend does not sends flow_id in response, which is necessary for the operations. Hence, kept disabled for now.

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

nmunet commented:
There are still some refactoring left with some e2e, WIP.

cloudbehl commented:
Thanks for PR. Mostly looks good! I have not tested it yet. 

Few questions:
1. Why do we have bucket selection first and flow Id second when we bucket is not even mandatory?
2. Why is bucket not a required parameter and what happens when value is not set?
3. Based on the bucket selected, can we select the zone automatically in which that bucket exists? 

Can we change `Flow Id` to `Name` and make it a first parameter?

nmunet commented:
jenkins test dashboard

nmunet commented:
jenkins test api

nmunet commented:
jenkins test make check

nmunet commented:
jenkins test make check arm64

nmunet commented:
> Thanks for PR. Mostly looks good! I have not tested it yet.
> 
> Few questions:
> 
> 1. Why do we have bucket selection first and flow Id second when we bucket is not even mandatory?
> 2. Why is bucket not a required parameter and what happens when value is not set?
> 3. Based on the bucket selected, can we select the zone automatically in which that bucket exists?
> 
> Can we change `Flow Id` to `Name` and make it a first parameter?

1) I have addressed this issue in my latest commit
2) we can only set bucket value if the sync policy is created with a bucket name, else not. And if we do not provide a bucket name it creates the sync policy/flow/pipe on the zonegroup level.
3) If we are talking about symmetrical flow then the zones could be multisite, in case of directional maybe we can select zone in which bucket exists.
4) Changed 'flow id' to 'name'

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-flow-modal/rgw-multisite-sync-flow-modal.component.html:
```diff
@@ -0,0 +1,121 @@
+              <input
+                class="form-control"
+                type="text"
+                placeholder="Flow Id..."
```
```suggestion
                placeholder="Flow Name..."
```

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-message.service.ts:
```diff
@@ -334,6 +334,14 @@ export class TaskMessageService {
+      this.commonOperations.delete,
+      (metadata) => {
+        return $localize`${
+          metadata.flow_ids.length > 1 ? 'selected Flow Ids' : `Flow Id '${metadata.flow_ids[0]}'`
```
```suggestion
          metadata.flow_ids.length > 1 ? 'selected Flow Names' : `Flow Name '${metadata.flow_ids[0]}'`
```

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-sync-policy-details/rgw-multisite-sync-policy-details.component.ts:
```diff
@@ -0,0 +1,228 @@
+      }
+    ];
+    this.directionalFlowCols = [
+      {
+        name: 'Source Zone',
+        prop: 'source_zone',
+        flexGrow: 1
+      },
+      {
+        name: 'Destination Zone',
+        prop: 'dest_zone',
+        flexGrow: 1
+      }
```
Why are we not showing name of directional flow?
```suggestion
      {
        name: 'Name',
        prop: 'id',
        flexGrow: 1
      },
      {
        name: 'Source Zone',
        prop: 'source_zone',
        flexGrow: 1
      },
      {
        name: 'Destination Zone',
        prop: 'dest_zone',
        flexGrow: 1
      }
```

nmunet replied:
we are not getting flow name from the API, there is already a bz opened for this issue.
That's why we are disabling the directional flow edit  delete functionality for now!

nmunet replied:
mentioned this in my PR description already!

cloudbehl replied:
Ah I missed it!

cloudbehl reviewed (approved):
LGTM

Some nits:

nmunet commented:
Updated PR with e2es and review suggestions!

nmunet commented:
jenkins test dashboard cephadm

nmunet commented:
jenkins test dashboard

nmunet commented:
jenkins test make check

nmunet commented:
jenkins test make check arm64

nmunet commented:
jenkins test dashboard cephadm

nmunet commented:
jenkins test dashboard

nmunet commented:
jenkins test api

nmunet commented:
jenkins test make check

nmunet commented:
jenkins test dashboard

nmunet commented:
jenkins test dashboard

nmunet commented:
jenkins test dashboard cephadm

nmunet commented:
jenkins test dashboard cephadm

nmunet commented:
jenkins test dashboard

nmunet commented:
jenkins test make check arm64

nmunet commented:
jenkins test dashboard
